evm/vm/base.py

```diff
@@ -66,19 +57,17 @@ def configure(cls,
                 )
         return type(name, (cls,), overrides)
 
-    _block = None
-    state_db = None
-
-    @property
-    def block(self):
-        if self._block is None:
-            raise AttributeError("No block property set")
-        return self._block
-
-    @block.setter
-    def block(self, value):
-        self._block = value
-        self.state_db = State(db=self.db, root_hash=value.header.state_root)
+    @contextmanager
+    def state_db(self, read_only=False):
+        state = State(db=self.db, root_hash=self.block.header.state_root)
+        yield state
+        if read_only:
+            # TODO: This is a bit of a hack; ideally we should raise an error whenever the
```

Comment:
I think I like the idea of implementing this as a wrapper around the `self.db` which raises errors on any write operations.  Might be a good *easy pickings* ticket to open.